File: gitlfs-server/src/main/java/ru/bozaro/gitlfs/server/PointerServlet.java
package ru.bozaro.gitlfs.server;

import ru.bozaro.gitlfs.common.Constants;
import ru.bozaro.gitlfs.common.JsonHelper;
import ru.bozaro.gitlfs.common.data.Error;
import ru.bozaro.gitlfs.common.data.*;
import ru.bozaro.gitlfs.server.internal.ObjectResponse;
import ru.bozaro.gitlfs.server.internal.ResponseWriter;

import javax.annotation.CheckForNull;
import javax.annotation.Nonnull;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.Arrays;
import java.util.Map;
import java.util.Objects;
import java.util.TreeMap;
import java.util.regex.Pattern;

/**
 * Servlet for pointer storage.
 * <p>
 * This servlet is entry point for git-lfs client.
 * <p>
 * Need to be mapped by path: objects/
 * <p>
 * Supported URL paths:
 * <p>
 * * POST: /
 * * POST: /batch
 * * GET:  /:oid
 *
 * @author Artem V. Navrotskiy
 */
public class PointerServlet extends HttpServlet {
  @Nonnull
  private static final Pattern PATTERN_OID = Pattern.compile("^/[0-9a-f]{64}$");
  @Nonnull
  private final PointerManager manager;
  @Nonnull
  private final AccessCheckerVisitor accessCheckerVisitor;

  /**
   * Create pointer manager for local ContentManager.
   *
   * @param manager         Content manager.
   * @param contentLocation Absolute or relative URL to ContentServlet.
   */
  public PointerServlet(@Nonnull ContentManager manager, @Nonnull String contentLocation) {
    this(new LocalPointerManager(manager, contentLocation));
  }

  public PointerServlet(@Nonnull PointerManager manager) {
    this.manager = manager;
    this.accessCheckerVisitor = new AccessCheckerVisitor(manager);
  }

  @Nonnull
  private static BatchItem[] filterLocations(@Nonnull BatchItem[] items, @Nonnull LocationFilter filter) throws IOException {
    final BatchItem[] result = new BatchItem[items.length];
    for (int i = 0; i < items.length; ++i) {
      if (items[i].getError() == null) {
        result[i] = filter.filter(items[i]);
      } else {
        result[i] = items[i];
      }
    }
    return result;
  }

  @Nonnull
  private static BatchItem filterDownload(@Nonnull BatchItem item) {
    if (item.getLinks().containsKey(LinkType.Download))
      return new BatchItem(item.getOid(), item.getSize(), filterLocation(item.getLinks(), LinkType.Download), null, null);

    return new BatchItem(item.getOid(), item.getSize(), null, null, new Error(HttpServletResponse.SC_NOT_FOUND, "Object not found"));
  }

  private static Map<LinkType, Link> filterLocation(@Nonnull Map<LinkType, Link> links, @Nonnull LinkType... linkTypes) {
    final Map<LinkType, Link> result = new TreeMap<>();
    for (LinkType linkType : linkTypes) {
      final Link link = links.get(linkType);
      if (link != null) result.put(linkType, link);
    }
    return result;
  }

  @Nonnull
  private static BatchItem filterUpload(@Nonnull BatchItem item) throws IOException {
    if (item.getLinks().containsKey(LinkType.Download))
      return new BatchItem(item.getOid(), item.getSize(), filterLocation(item.getLinks(), LinkType.Verify), null, null);

    if (item.getLinks().containsKey(LinkType.Upload))
      return new BatchItem(item.getOid(), item.getSize(), filterLocation(item.getLinks(), LinkType.Upload, LinkType.Verify), null, null);

    throw new IOException("Upload link not found");
  }

  @Override
  protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
    try {
      if ((req.getPathInfo() != null) && PATTERN_OID.matcher(req.getPathInfo()).matches()) {
        processObjectGet(req, req.getPathInfo().substring(1)).write(resp);
        return;
      }
    } catch (ServerError e) {
      PointerServlet.sendError(resp, e);
      return;
    }
    super.doGet(req, resp);
  }

  @Override
  protected void doPost(@Nonnull HttpServletRequest req, @Nonnull HttpServletResponse resp) throws ServletException, IOException {
    try {
      checkMimeTypes(req);

      if (req.getPathInfo() == null) {
        processObjectPost(req).write(resp);
        return;
      }
      if ("/batch".equals(req.getPathInfo())) {
        processBatchPost(req).write(resp);
        return;
      }
    } catch (ServerError e) {
      sendError(resp, e);
      return;
    }
    super.doPost(req, resp);
  }

  public static void checkMimeTypes(@Nonnull HttpServletRequest request) throws ServerError {
    checkMimeType(request.getContentType());
    checkMimeType(request.getHeader(Constants.HEADER_ACCEPT));
  }

  @Nonnull
  private ResponseWriter processObjectPost(@Nonnull HttpServletRequest req) throws ServerError, IOException {
    final URI selfUrl = getSelfUrl(req);
    final PointerManager.Locator locator = manager.checkUploadAccess(req, selfUrl);
    final Meta meta = JsonHelper.mapper.readValue(req.getInputStream(), Meta.class);
    final BatchItem location = getLocation(locator, meta);
    final Error error = location.getError();
    if (error != null) {
      throw new ServerError(error.getCode(), error.getMessage(), null);
    }
    final Map<LinkType, Link> links = new TreeMap<>(location.getLinks());
    links.put(LinkType.Self, new Link(selfUrl, null, null));
    if (links.containsKey(LinkType.Download)) {
      return new ObjectResponse(HttpServletResponse.SC_OK, new ObjectRes(location.getOid(), location.getSize(), addSelfLink(req, links)));
    }
    if (links.containsKey(LinkType.Upload)) {
      return new ObjectResponse(HttpServletResponse.SC_ACCEPTED, new ObjectRes(location.getOid(), location.getSize(), addSelfLink(req, links)));
    }
    throw new ServerError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Invalid locations list", null);
  }

  @Nonnull
  private ResponseWriter processBatchPost(@Nonnull HttpServletRequest req) throws ServerError, IOException {
    final BatchReq batchReq = JsonHelper.mapper.readValue(req.getInputStream(), BatchReq.class);
    final PointerManager.Locator locator = batchReq.getOperation().visit(accessCheckerVisitor).checkAccess(req, getSelfUrl(req));
    final BatchItem[] locations = getLocations(locator, batchReq.getObjects().toArray(new Meta[0]));
    return new ObjectResponse(HttpServletResponse.SC_OK, new BatchRes(Arrays.asList(locations)));
  }

  public static void sendError(@Nonnull HttpServletResponse resp, @Nonnull ServerError e) throws IOException {
    resp.setStatus(e.getStatusCode());
    resp.setContentType(Constants.MIME_LFS_JSON);
    JsonHelper.mapper.writeValue(resp.getOutputStream(), new Error(e.getStatusCode(), e.getMessage()));
  }

  private static void checkMimeType(@CheckForNull String contentType) throws ServerError {
    String actualType = contentType;
    if (actualType != null) {
      int separator = actualType.indexOf(';');
      if (separator >= 0) {
        while (separator > 1 && actualType.charAt(separator - 1) == ' ') {
          separator--;
        }
        actualType = actualType.substring(0, separator);
      }
    }
    if (!Constants.MIME_LFS_JSON.equals(actualType)) {
      throw new ServerError(HttpServletResponse.SC_NOT_ACCEPTABLE, "Not Acceptable", null);
    }
  }

  @Nonnull
  protected URI getSelfUrl(@Nonnull HttpServletRequest req) {
    try {
      return new URI(req.getScheme(), null, req.getServerName(), req.getServerPort(), req.getServletPath(), null, null);
    } catch (URISyntaxException e) {
      throw new IllegalStateException("Can't create request URL", e);
    }
  }

  @Nonnull
  private BatchItem getLocation(@Nonnull PointerManager.Locator locator, @Nonnull Meta meta) throws IOException, ServerError {
    return getLocations(locator, new Meta[]{meta})[0];
  }

  private static Map<LinkType, Link> addSelfLink(@Nonnull HttpServletRequest req, @Nonnull Map<LinkType, Link> links) {
    final Map<LinkType, Link> result = new TreeMap<>(links);
    result.put(LinkType.Self, new Link(URI.create(String.valueOf(req.getRequestURL())), null, null));
    return result;
  }

  @Nonnull
  private BatchItem[] getLocations(@Nonnull PointerManager.Locator locator, @Nonnull Meta[] metas) throws ServerError, IOException {
    final BatchItem[] locations = locator.getLocations(metas);
    // Invalid locations list.
    if (locations.length != metas.length) {
      throw new ServerError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Unexpected locations array size", null);
    }
    for (int i = 0; i < locations.length; ++i) {
      if (!Objects.equals(metas[i].getOid(), locations[i].getOid())) {
        throw new ServerError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Metadata mismatch", null);
      }
    }
    return locations;
  }

  @Nonnull
  private ResponseWriter processObjectGet(@Nonnull HttpServletRequest req, @Nonnull String oid) throws ServerError, IOException {
    final PointerManager.Locator locator = manager.checkDownloadAccess(req, getSelfUrl(req));
    final BatchItem location = getLocation(locator, new Meta(oid, -1));
    // Return error information.
    final Error error = location.getError();
    if (error != null) {
      throw new ServerError(error.getCode(), error.getMessage(), null);
    }
    if (location.getLinks().containsKey(LinkType.Download)) {
      return new ObjectResponse(HttpServletResponse.SC_OK, new ObjectRes(location.getOid(), location.getSize(), addSelfLink(req, location.getLinks())));
    }
    throw new ServerError(HttpServletResponse.SC_NOT_FOUND, "Object not found", null);
  }

  @FunctionalInterface
  protected interface AccessChecker {
    @Nonnull
    PointerManager.Locator checkAccess(@Nonnull HttpServletRequest request, @Nonnull URI selfUrl) throws IOException, ForbiddenError, UnauthorizedError;
  }

  @FunctionalInterface
  protected interface LocationFilter {
    @Nonnull
    BatchItem filter(@Nonnull BatchItem item) throws IOException;
  }

  private static class AccessCheckerVisitor implements Operation.Visitor<AccessChecker> {
    @Nonnull
    private final PointerManager manager;

    public AccessCheckerVisitor(@Nonnull PointerManager manager) {
      this.manager = manager;
    }

    @Override
    public AccessChecker visitDownload() {
      return wrapChecker(manager::checkDownloadAccess, PointerServlet::filterDownload);
    }

    @Override
    public AccessChecker visitUpload() {
      return wrapChecker(manager::checkUploadAccess, PointerServlet::filterUpload);
    }

    private AccessChecker wrapChecker(@Nonnull AccessChecker checker, @Nonnull LocationFilter filter) {
      return (request, selfUrl) -> {
        PointerManager.Locator locator = checker.checkAccess(request, selfUrl);
        return metas -> filterLocations(locator.getLocations(metas), filter);
      };
    }
  }
}

File: gitlfs-server/src/main/kotlin/ru/bozaro/gitlfs/server/PointerServlet.kt
package ru.bozaro.gitlfs.server

import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServlet
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import ru.bozaro.gitlfs.common.Constants
import ru.bozaro.gitlfs.common.JsonHelper
import ru.bozaro.gitlfs.common.data.*
import ru.bozaro.gitlfs.server.PointerServlet.AccessChecker
import ru.bozaro.gitlfs.server.internal.ObjectResponse
import ru.bozaro.gitlfs.server.internal.ResponseWriter
import java.io.IOException
import java.net.URI
import java.net.URISyntaxException
import java.util.*
import java.util.regex.Pattern

/**
 * Servlet for pointer storage.
 *
 *
 * This servlet is entry point for git-lfs client.
 *
 *
 * Need to be mapped by path: objects/
 *
 *
 * Supported URL paths:
 *
 *
 * * POST: /
 * * POST: /batch
 * * GET:  /:oid
 *
 * @author Artem V. Navrotskiy
 */
class PointerServlet(private val manager: PointerManager) : HttpServlet() {
    private val accessCheckerVisitor: AccessCheckerVisitor = AccessCheckerVisitor(manager)

    /**
     * Create pointer manager for local ContentManager.
     *
     * @param manager         Content manager.
     * @param contentLocation Absolute or relative URL to ContentServlet.
     */
    constructor(manager: ContentManager, contentLocation: String) : this(
        LocalPointerManager(
            manager,
            contentLocation
        )
    )

    @Throws(ServletException::class, IOException::class)
    override fun doGet(req: HttpServletRequest, resp: HttpServletResponse) {
        try {
            if (req.pathInfo != null && PATTERN_OID.matcher(req.pathInfo).matches()) {
                processObjectGet(req, req.pathInfo.substring(1)).write(resp)
                return
            }
        } catch (e: ServerError) {
            sendError(resp, e)
            return
        }
        super.doGet(req, resp)
    }

    @Throws(ServletException::class, IOException::class)
    override fun doPost(req: HttpServletRequest, resp: HttpServletResponse) {
        try {
            checkMimeTypes(req)
            if (req.pathInfo == null) {
                processObjectPost(req).write(resp)
                return
            }
            if ("/batch" == req.pathInfo) {
                processBatchPost(req).write(resp)
                return
            }
        } catch (e: ServerError) {
            sendError(resp, e)
            return
        }
        super.doPost(req, resp)
    }

    @Throws(ServerError::class, IOException::class)
    private fun processObjectPost(req: HttpServletRequest): ResponseWriter {
        val selfUrl = getSelfUrl(req)
        val locator = manager.checkUploadAccess(req, selfUrl)
        val meta = JsonHelper.mapper.readValue(req.inputStream, Meta::class.java)
        val location = getLocation(locator, meta)
        val error = location.error
        if (error != null) {
            throw ServerError(error.code, error.message, null)
        }
        val links: MutableMap<LinkType, Link> = TreeMap(location.links)
        links[LinkType.Self] = Link(selfUrl, null, null)
        if (links.containsKey(LinkType.Download)) {
            return ObjectResponse(
                HttpServletResponse.SC_OK,
                ObjectRes(location.oid, location.size, addSelfLink(req, links))
            )
        }
        if (links.containsKey(LinkType.Upload)) {
            return ObjectResponse(
                HttpServletResponse.SC_ACCEPTED,
                ObjectRes(location.oid, location.size, addSelfLink(req, links))
            )
        }
        throw ServerError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Invalid locations list", null)
    }

    @Throws(ServerError::class, IOException::class)
    private fun processBatchPost(req: HttpServletRequest): ResponseWriter {
        val batchReq = JsonHelper.mapper.readValue(req.inputStream, BatchReq::class.java)
        val locator = batchReq.operation.visit(accessCheckerVisitor).checkAccess(req, getSelfUrl(req))
        val locations = getLocations(locator, batchReq.objects.toTypedArray())
        return ObjectResponse(HttpServletResponse.SC_OK, BatchRes(listOf(*locations)))
    }

    private fun getSelfUrl(req: HttpServletRequest): URI {
        return try {
            URI(req.scheme, null, req.serverName, req.serverPort, req.servletPath, null, null)
        } catch (e: URISyntaxException) {
            throw IllegalStateException("Can't create request URL", e)
        }
    }

    @Throws(IOException::class, ServerError::class)
    private fun getLocation(locator: PointerManager.Locator, meta: Meta): BatchItem {
        return getLocations(locator, arrayOf(meta))[0]
    }

    @Throws(ServerError::class, IOException::class)
    private fun getLocations(
        locator: PointerManager.Locator,
        metas: Array<Meta>
    ): Array<BatchItem> {
        val locations = locator.getLocations(metas)
        // Invalid locations list.
        if (locations.size != metas.size) {
            throw ServerError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Unexpected locations array size", null)
        }
        for (i in locations.indices) {
            if (metas[i].oid != locations[i].oid) {
                throw ServerError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Metadata mismatch", null)
            }
        }
        return locations
    }

    @Throws(ServerError::class, IOException::class)
    private fun processObjectGet(req: HttpServletRequest, oid: String): ResponseWriter {
        val locator = manager.checkDownloadAccess(req, getSelfUrl(req))
        val location = getLocation(locator, Meta(oid, -1))
        // Return error information.
        val error = location.error
        if (error != null) {
            throw ServerError(error.code, error.message, null)
        }
        if (location.links.containsKey(LinkType.Download)) {
            return ObjectResponse(
                HttpServletResponse.SC_OK,
                ObjectRes(location.oid, location.size, addSelfLink(req, location.links))
            )
        }
        throw ServerError(HttpServletResponse.SC_NOT_FOUND, "Object not found", null)
    }

    private fun interface AccessChecker {
        @Throws(IOException::class, ForbiddenError::class, UnauthorizedError::class)
        fun checkAccess(request: HttpServletRequest, selfUrl: URI): PointerManager.Locator
    }

    private fun interface LocationFilter {
        @Throws(IOException::class)
        fun filter(item: BatchItem): BatchItem
    }

    private class AccessCheckerVisitor(private val manager: PointerManager) : Operation.Visitor<AccessChecker> {
        override fun visitDownload(): AccessChecker {
            return wrapChecker({ request: HttpServletRequest, selfUrl: URI ->
                manager.checkDownloadAccess(
                    request,
                    selfUrl
                )
            }, { item: BatchItem -> filterDownload(item) })
        }

        override fun visitUpload(): AccessChecker {
            return wrapChecker({ request: HttpServletRequest, selfUrl: URI ->
                manager.checkUploadAccess(
                    request,
                    selfUrl
                )
            }, { item: BatchItem -> filterUpload(item) })
        }

        private fun wrapChecker(checker: AccessChecker, filter: LocationFilter): AccessChecker {
            return AccessChecker { request: HttpServletRequest, selfUrl: URI ->
                val locator = checker.checkAccess(request, selfUrl)
                PointerManager.Locator { metas: Array<Meta> -> filterLocations(locator.getLocations(metas), filter) }
            }
        }
    }

    companion object {
        val PATTERN_OID: Pattern = Pattern.compile("^/[0-9a-f]{64}$")

        @Throws(IOException::class)
        private fun filterLocations(
            items: Array<BatchItem>,
            filter: LocationFilter
        ): Array<BatchItem> {
            return items.map {
                if (it.error == null) {
                    filter.filter(it)
                } else {
                    it
                }
            }.toTypedArray()
        }

        private fun filterDownload(item: BatchItem): BatchItem {
            return if (item.links.containsKey(LinkType.Download)) BatchItem(
                item.oid,
                item.size,
                filterLocation(item.links, LinkType.Download),
                null,
                null
            ) else BatchItem(
                item.oid,
                item.size,
                null,
                null,
                Error(HttpServletResponse.SC_NOT_FOUND, "Object not found")
            )
        }

        private fun filterLocation(
            links: Map<LinkType, Link>,
            vararg linkTypes: LinkType
        ): Map<LinkType, Link> {
            val result = TreeMap<LinkType, Link>()
            for (linkType in linkTypes) {
                val link = links[linkType]
                if (link != null) result[linkType] = link
            }
            return result
        }

        @Throws(IOException::class)
        private fun filterUpload(item: BatchItem): BatchItem {
            if (item.links.containsKey(LinkType.Download)) return BatchItem(
                item.oid,
                item.size,
                filterLocation(item.links, LinkType.Verify),
                null,
                null
            )
            if (item.links.containsKey(LinkType.Upload)) return BatchItem(
                item.oid,
                item.size,
                filterLocation(item.links, LinkType.Upload, LinkType.Verify),
                null,
                null
            )
            throw IOException("Upload link not found")
        }

        @Throws(ServerError::class)
        fun checkMimeTypes(request: HttpServletRequest) {
            checkMimeType(request.contentType)
            checkMimeType(request.getHeader(Constants.HEADER_ACCEPT))
        }

        @Throws(IOException::class)
        fun sendError(resp: HttpServletResponse, e: ServerError) {
            resp.status = e.statusCode
            resp.contentType = Constants.MIME_LFS_JSON
            JsonHelper.mapper.writeValue(resp.outputStream, Error(e.statusCode, e.message))
        }

        @Throws(ServerError::class)
        private fun checkMimeType(contentType: String?) {
            var actualType = contentType
            if (actualType != null) {
                var separator = actualType.indexOf(';')
                if (separator >= 0) {
                    while (separator > 1 && actualType[separator - 1] == ' ') {
                        separator--
                    }
                    actualType = actualType.substring(0, separator)
                }
            }
            if (Constants.MIME_LFS_JSON != actualType) {
                throw ServerError(HttpServletResponse.SC_NOT_ACCEPTABLE, "Not Acceptable", null)
            }
        }

        private fun addSelfLink(
            req: HttpServletRequest,
            links: Map<LinkType, Link>
        ): Map<LinkType, Link> {
            val result: MutableMap<LinkType, Link> = TreeMap(links)
            result[LinkType.Self] = Link(URI.create(req.requestURL.toString()), null, null)
            return result
        }
    }

}
